File: backend/internal/modules/dispatch/domain/job.go
package domain

import (
	"errors"
	"time"
)

// JobStatus is where a delivery is in its own life.
//
// Deliberately not the order's status. An order is the customer's view of the
// thing; a job is the platform's view of getting it moved, and they diverge:
// an order sitting at `ready` may have been offered to four partners and
// declined by three, none of which is a fact the customer needs.
type JobStatus string

// The job lifecycle.
const (
	// JobWaiting is a job nobody holds. It appears in every eligible partner's
	// feed until somebody takes it.
	JobWaiting JobStatus = "waiting"
	// JobOffered is a job put to one partner, with a clock running
	// (dispatch.assignment_timeout).
	JobOffered JobStatus = "offered"
	// JobAssigned is a partner having accepted it.
	JobAssigned JobStatus = "assigned"
	// JobCollected is the rider having the goods.
	JobCollected JobStatus = "collected"
	// JobDelivered is done, and terminal.
	JobDelivered JobStatus = "delivered"
	// JobFailed is a delivery that could not be completed, and terminal.
	JobFailed JobStatus = "failed"
	// JobCancelled is the order behind it having gone away, and terminal.
	JobCancelled JobStatus = "cancelled"
)

// IsTerminal reports whether a job has finished.
func (s JobStatus) IsTerminal() bool {
	return s == JobDelivered || s == JobFailed || s == JobCancelled
}

// IsLive reports whether a job is still going.
func (s JobStatus) IsLive() bool { return !s.IsTerminal() && s != "" }

// The rules a job holds itself to.
var (
	ErrNoOrder          = errors.New("a job must be about an order")
	ErrJobIllegalMove   = errors.New("a job cannot move that way")
	ErrJobNotYours      = errors.New("that job is not yours")
	ErrJobFinished      = errors.New("that job has already finished")
	ErrJobUnassigned    = errors.New("that job has nobody on it")
	ErrOfferExpired     = errors.New("that offer has expired")
	ErrNegativeDistance = errors.New("a distance cannot be negative")
	ErrNoFailureReason  = errors.New("a failed delivery needs a reason")
)

// Place is one end of a delivery.
type Place struct {
	Name       string
	Phone      string
	SingleLine string
	Lat        float64
	Lng        float64
}

// Placement is where a job is, in administrative terms.
//
// Carried on the job because the settings that govern it — the offer timeout,
// the candidate radius — are per-area (D2), and the sweeper that re-offers a
// job later has only the job to go on.
type Placement struct {
	AreaCode     string
	DistrictCode string
	DivisionCode string
}

// Job is one delivery to be made.
type Job struct {
	ID      string
	OrderID string
	// Code is the order's short reference, carried so a rider and a shopkeeper
	// can say the same six characters to each other.
	Code string

	Status JobStatus
	// PartnerID is who holds it — the partner it is offered to, or the one who
	// accepted. Empty while waiting.
	PartnerID string

	Pickup      Place
	Destination Place
	DistanceM   float64
	Band        Band
	Placement   Placement

	// PassedBy is the last partner who had this job and did not take it —
	// declined it, or let the offer run out. The next round skips them, so a
	// rider who has already said no is not asked the same question twice
	// while somebody else is standing by. Cleared on the next offer.
	PassedBy string

	// Reason is why a job ended the way it did — the order was cancelled, the
	// customer was not in. Kept because a failed delivery is the start of a
	// support conversation, and "failed" on its own starts it badly.
	Reason string
	// Attempts is how many partners have been offered this and not taken it.
	// An order nobody wants is an operational fact somebody has to see.
	Attempts int
	// OfferedAt and OfferExpiresAt bound the current offer.
	OfferedAt      time.Time
	OfferExpiresAt time.Time

	CreatedAt time.Time
	UpdatedAt time.Time
}

// NewJob opens a delivery for an order that is ready to collect.
func NewJob(
	id, orderID, code string,
	pickup, destination Place,
	distanceM float64, band Band, placement Placement,
	now time.Time,
) (Job, error) {
	switch {
	case orderID == "":
		return Job{}, ErrNoOrder
	case distanceM < 0:
		return Job{}, ErrNegativeDistance
	}
	return Job{
		ID: id, OrderID: orderID, Code: code,
		Status:      JobWaiting,
		Pickup:      pickup,
		Destination: destination,
		DistanceM:   distanceM,
		Band:        band,
		Placement:   placement,
		CreatedAt:   now, UpdatedAt: now,
	}, nil
}

// Offer puts a job to one partner, with a clock running.
//
// The clock is the point. A partner who opens the app, sees an offer and puts
// the phone in their pocket must not hold a customer's dinner hostage; the
// offer lapses and the job goes back into everybody's feed.
func (j *Job) Offer(partnerID string, at time.Time, timeout time.Duration) error {
	if j.Status.IsTerminal() {
		return ErrJobFinished
	}
	if j.Status != JobWaiting && !j.OfferLapsed(at) {
		return ErrJobIllegalMove
	}
	j.Status = JobOffered
	j.PartnerID = partnerID
	j.PassedBy = ""
	j.OfferedAt = at
	j.OfferExpiresAt = at.Add(timeout)
	j.Attempts++
	j.UpdatedAt = at
	return nil
}

// OfferLapsed reports whether the current offer has run out.
func (j Job) OfferLapsed(at time.Time) bool {
	return j.Status == JobOffered && !at.Before(j.OfferExpiresAt)
}

// Accept is a partner taking the job they were offered.
func (j *Job) Accept(partnerID string, at time.Time) error {
	if err := j.heldBy(partnerID); err != nil {
		return err
	}
	if j.Status != JobOffered {
		return ErrJobIllegalMove
	}
	if j.OfferLapsed(at) {
		return ErrOfferExpired
	}
	j.Status = JobAssigned
	j.OfferExpiresAt = time.Time{}
	j.UpdatedAt = at
	return nil
}

// Decline is a partner passing. The job goes back to waiting and the next
// round finds somebody else.
func (j *Job) Decline(partnerID string, at time.Time) error {
	if err := j.heldBy(partnerID); err != nil {
		return err
	}
	if j.Status != JobOffered {
		return ErrJobIllegalMove
	}
	return j.release(at)
}

// Lapse expires an offer nobody answered.
func (j *Job) Lapse(at time.Time) error {
	if !j.OfferLapsed(at) {
		return ErrJobIllegalMove
	}
	return j.release(at)
}

// release returns a job to the pool.
func (j *Job) release(at time.Time) error {
	j.PassedBy = j.PartnerID
	j.Status = JobWaiting
	j.PartnerID = ""
	j.OfferedAt = time.Time{}
	j.OfferExpiresAt = time.Time{}
	j.UpdatedAt = at
	return nil
}

// Collect is the rider having the goods.
func (j *Job) Collect(partnerID string, at time.Time) error {
	return j.move(partnerID, JobAssigned, JobCollected, at)
}

// Deliver is the rider having handed them over.
func (j *Job) Deliver(partnerID string, at time.Time) error {
	return j.move(partnerID, JobCollected, JobDelivered, at)
}

// Fail is a rider giving up on a delivery — nobody at the address, an accident.
//
// What that means depends on whether they have the goods. Before collection the
// food is still on the shop's counter, so there is nothing to fail: the job
// goes back on the board for somebody else and the customer's order carries on.
// After collection the rider is holding somebody's dinner and the delivery
// really has failed.
func (j *Job) Fail(partnerID, reason string, at time.Time) error {
	if err := j.heldBy(partnerID); err != nil {
		return err
	}
	if reason == "" {
		// A rider abandoning a delivery owes the customer and the shop an
		// explanation. It is the only thing either of them will have to go on.
		return ErrNoFailureReason
	}
	switch j.Status {
	case JobAssigned:
		j.Reason = reason
		return j.release(at)
	case JobCollected:
		j.Status = JobFailed
		j.Reason = reason
		j.UpdatedAt = at
		return nil
	default:
		return ErrJobIllegalMove
	}
}

// Cancel takes a job off the board because the order behind it went away.
//
// No partner id: this is the platform acting, not a person. A rider holding a
// cancelled job finds out because it leaves their list.
func (j *Job) Cancel(reason string, at time.Time) error {
	if j.Status.IsTerminal() {
		return ErrJobFinished
	}
	j.Status = JobCancelled
	j.Reason = reason
	j.UpdatedAt = at
	return nil
}

// move applies a straightforward transition by the partner holding the job.
func (j *Job) move(partnerID string, from, to JobStatus, at time.Time) error {
	if err := j.heldBy(partnerID); err != nil {
		return err
	}
	if j.Status != from {
		return ErrJobIllegalMove
	}
	j.Status = to
	j.UpdatedAt = at
	return nil
}

// heldBy checks that this partner is the one who has the job.
//
// The acceptance criterion "two partners cannot be assigned the same order" has
// two halves. This is the one in the domain: a partner who is not holding a job
// cannot move it. The other half is a compare-and-set in the repository, which
// is what stops two partners accepting the same waiting job in the same
// millisecond.
func (j Job) heldBy(partnerID string) error {
	switch {
	case j.Status.IsTerminal():
		return ErrJobFinished
	case j.PartnerID == "":
		return ErrJobUnassigned
	case j.PartnerID != partnerID:
		return ErrJobNotYours
	}
	return nil
}

File: frontend/partner/lib/src/api/models/job.dart
import 'package:flutter/foundation.dart';
import 'package:goklay_core/goklay_core.dart';

import 'partner.dart';

/// One end of a delivery.
@immutable
class DeliveryPlace {
  /// Creates a place.
  const DeliveryPlace({
    required this.name,
    required this.phone,
    required this.singleLine,
    required this.lat,
    required this.lng,
  });

  /// Reads the `DeliveryPlace` object.
  factory DeliveryPlace.fromJson(Map<String, Object?> json) => DeliveryPlace(
    name: readString(json, 'name'),
    phone: readString(json, 'phone'),
    singleLine: readString(json, 'single_line'),
    lat: readDouble(json, 'lat'),
    lng: readDouble(json, 'lng'),
  );

  /// Who is there.
  final String name;

  /// A number to call.
  final String phone;

  /// The address on one line, composed by the server.
  final String singleLine;

  /// Where to go.
  final double lat;

  /// Where to go.
  final double lng;
}

/// One delivery, offered or held.
///
/// [band] is D4's distance band and [distance] is the pickup-to-door figure
/// already rendered as a string. The app never works out a distance: the
/// server has the coordinates, the ladder and the rounding, and a second
/// calculation here would disagree with the one that decided the job was
/// offered at all.
@immutable
class DeliveryJob {
  /// Creates a job.
  const DeliveryJob({
    required this.id,
    required this.orderId,
    required this.code,
    required this.status,
    required this.statusLabel,
    required this.live,
    required this.band,
    required this.bandLabel,
    required this.distance,
    required this.toPickup,
    required this.pickup,
    required this.destination,
    required this.reason,
    required this.secondsLeft,
  });

  /// Reads the `DeliveryJob` object.
  factory DeliveryJob.fromJson(Map<String, Object?> json) => DeliveryJob(
    id: readString(json, 'id'),
    orderId: readString(json, 'order_id'),
    code: readString(json, 'code'),
    status: readString(json, 'status'),
    statusLabel: readString(json, 'status_label'),
    live: readBool(json, 'live'),
    band: readString(json, 'band'),
    bandLabel: readString(json, 'band_label'),
    distance: readString(json, 'distance'),
    toPickup: readString(json, 'to_pickup'),
    pickup: DeliveryPlace.fromJson(readObject(json, 'pickup')),
    destination: DeliveryPlace.fromJson(readObject(json, 'destination')),
    reason: readString(json, 'reason'),
    secondsLeft: readInt(json, 'seconds_left'),
  );

  /// The job's id.
  final String id;

  /// The order behind it.
  final String orderId;

  /// The order's short code, so a rider and a shopkeeper say the same six
  /// characters to each other.
  final String code;

  /// `waiting`, `offered`, `assigned`, `collected`, `delivered`, `failed` or
  /// `cancelled`.
  final String status;

  /// That state as a sentence.
  final String statusLabel;

  /// Whether it is still going.
  final bool live;

  /// `short`, `long` or `beyond`.
  final String band;

  /// That band as a sentence.
  final String bandLabel;

  /// Pickup to door, rendered.
  final String distance;

  /// How far the rider is from the counter, rendered.
  final String toPickup;

  /// Where to collect.
  final DeliveryPlace pickup;

  /// Where to deliver.
  final DeliveryPlace destination;

  /// Why it failed, where it did.
  final String reason;

  /// How long an offer has left. The server's number; the screen counts it
  /// down for the look of the thing and re-reads the feed rather than
  /// deciding for itself that an offer has lapsed.
  final int secondsLeft;

  /// A job that has been offered and not yet answered.
  static const String offered = 'offered';

  /// A job this rider holds but has not collected.
  static const String assigned = 'assigned';

  /// A job in the rider's bag.
  static const String collected = 'collected';

  /// Whether the rider may accept or decline it.
  bool get isOffer => status == offered;

  /// Whether the next thing to do is collect it.
  bool get awaitsCollection => status == assigned;

  /// Whether the next thing to do is hand it over.
  bool get awaitsDelivery => status == collected;
}

/// A page of the rider's own deliveries.
@immutable
class DeliveryJobList {
  /// Creates a page.
  const DeliveryJobList({required this.jobs, required this.total});

  /// Reads the `DeliveryJobList` response.
  factory DeliveryJobList.fromJson(Map<String, Object?> json) =>
      DeliveryJobList(
        jobs: readList(json, 'jobs', DeliveryJob.fromJson),
        total: readInt(json, 'total'),
      );

  /// This page's jobs.
  final List<DeliveryJob> jobs;

  /// How many matched before paging.
  final int total;

  /// Whether the rider holds nothing.
  bool get isEmpty => jobs.isEmpty;
}

/// What the rider should be looking at right now (ALG-08).
///
/// The feed is bounded, nearest-pickup first, and already filtered by the
/// rider's own distance choice. A client that filtered a global list by its
/// own radius would be deciding visibility — which 2.9 forbids, and which
/// would need a copy of `dispatch.partner_radius` to do at all.
///
/// An empty feed carries a [reason] and a [notice]. A rider who forgot to go
/// on shift is told that, rather than shown nothing and left to wonder.
@immutable
class PartnerFeed {
  /// Creates a feed.
  const PartnerFeed({
    required this.partner,
    required this.jobs,
    required this.reason,
    required this.notice,
  });

  /// Reads the `PartnerFeed` response.
  factory PartnerFeed.fromJson(Map<String, Object?> json) => PartnerFeed(
    partner: DeliveryPartner.fromJson(readObject(json, 'partner')),
    jobs: readList(json, 'jobs', DeliveryJob.fromJson),
    reason: readOptionalString(json, 'reason'),
    notice: readString(json, 'notice'),
  );

  /// The rider this feed is for.
  final DeliveryPartner partner;

  /// What they are being offered.
  final List<DeliveryJob> jobs;

  /// `offline`, `at_capacity` or `nothing_nearby` when the list is empty.
  /// A code to branch on, never to print.
  final String? reason;

  /// The sentence to show, composed by the server.
  final String notice;

  /// Whether there is anything to take.
  bool get isEmpty => jobs.isEmpty;
}
